Locate type at point

andreypopp opened the pull request:
Related to #885

Modifies `Locate.from_string` to accept `kind : [ Any | Type ]` argument:

- In case `Any` is passed then the original behaviour is used.
- In case `Type` is passed we try to lookup only types from environment.

Todo:

- [x] Tests
- [x] There's a short circuit in `inspect_pattern` which doesn't hold if we ask for type

andreypopp commented on src/analysis/locate.ml:
```diff
@@ -621,20 +622,23 @@ module Env_lookup : sig
+    | `Any, Label _       -> [ `Labels; `Mod ]
+    | `Any, Constructor _ -> [ `Constr; `Mod ]
+    | `Any, Module_path   -> [ `Mod ]
+    | `Type, (Type | Module_type | Expr | Patt | Unknown) -> [ `Type ]
```
Not sure if we need to consider others than `Type` here.

trefis commented on src/analysis/locate.mli:
```diff
@@ -33,6 +33,7 @@ val from_string
   -> env:Env.t
   -> local_defs:Mtyper.typedtree
   -> pos:Lexing.position
+  -> [ `Any | `Type ]
```
I think it doesn't hurt to make this more precise, so I'd suggest having:
```ocaml
    -> ?namespace:[ `Type | `Module | `Value | ... ]
```

Instead of this.

And at that point, exposing a proper (named) type might also make sense?

trefis commented on src/analysis/locate.ml:
```diff
@@ -610,6 +610,7 @@ module Env_lookup : sig
+    -> [ `Any | `Type ]
     -> Longident.t
     -> Env.t
     -> (Path.t * Namespaced_path.t * Location.t) option
```
I think the API would be slightly nicer if it was changed to:
```ocaml
val namespaces_of_context : Context.t -> namespace list

val in_namespaces : namespace list -> Longident.t -> Env.t -> (Path.t * Namespaced_path.t * Location.t) option
```

andreypopp commented:
@trefis thanks for the review!

- e238e92
  I've added `type namespace` and exposed it in `Locate` module so consumers now can ask in which namespaces to locate the id. If no namespaces list is provided then context is being used to infer the list of namespaces.

- b58d33a
  Then I moved the "short circuit" code for patterns which leads to `At_origin` case to `Env_lookup` so it only does that when looking up in `Values` namespace. I hope this is correct (all tests pass) but I might have missed something (?) — not very knowledgeable about typedtree structure.

- 1e09a42
  Lastly I've added tests for LSP frontend for `textDocument/definition` (just a smoke test for now as most of the logic is tested in original merlin protocol tests) and `textDocument/typeDefinition` (this tests more cases of interested, e.g. when the same named value binding is present in environment).

I didn't modify original merlin protocol and thus this functionality is still exposed via LSP only. If one would want to do that though I think something like:
```
% ocamlmerlin locate -namespaces types,value ...
```
would work.

andreypopp commented on src/analysis/locate.ml:
```diff
@@ -320,7 +320,7 @@ module Context = struct
     | Module_path
     | Module_type
-    | Patt
+    | Patt of Typedtree.pattern_desc
```
Need that as we are inspecting the `let [x] = ...` case later now.

andreypopp commented on src/analysis/locate.ml:
```diff
@@ -608,64 +633,83 @@ let recover _ =
-          | `Labels ->
+            raise (Found (At (path, ns_path, typ_decl.Types.type_loc)))
+          | Values ->
+            begin match ctxt with
```
this logic was previously in `inspect_context`

andreypopp commented on src/analysis/locate.ml:
```diff
@@ -755,14 +806,6 @@ let inspect_pattern ~pos ~lid p =
     (fun fmt -> Format.fprintf fmt "current pattern is: %a"
                   (Printtyped.pattern 0) p);
   match p.pat_desc with
```
Moved to `Env_lookup`

andreypopp commented on src/analysis/locate.ml:
```diff
@@ -862,7 +906,16 @@ let get_doc ~config ~env ~local_defs ~comments ~pos =
+        begin match from_longident ~config ~pos ~env ~lazy_trie ctxt `MLI lid
+        with
+        | `At_origin ->
+          `Found ({ Location. loc_start=pos; loc_end=pos ; loc_ghost=true }, None)
```
If we are at origin — use the current location (same behaviour as `None` is returned from `inspect_context` before.

trefis commented:
> I didn't modify original merlin protocol and thus this functionality is still exposed via LSP only.

Me neither, I'll do that in a follow up PR once this is merged.

andreypopp commented:
> I took the liberty of rewriting/reorganizing some parts of your PR, I hope you don't mind

Sure!

> I think everything should work as expected, but I didn't run your tests.

It broke tests for some reason, I'll look into it. I'll push a commit which adds LSP tests to CI as a separate job.

trefis commented:
Actually, *I* broke the build on 4.02, I'll push a fix for that.

trefis commented:
The 4.02 build should now be fixed.
I'll let you have a look at the tests, ping me if there's something that looks weird to you!

trefis commented:
@andreypopp: can we merge this? Or do you want to rebase it and/or update the lsp tests?

andreypopp commented:
The last commits broke lsp tests I need to take a look at it.

On Mon, 25 Feb 2019 at 15:40, Thomas Refis <notifications@github.com> wrote:

> @andreypopp <https://github.com/andreypopp>: can we merge this? Or do you
> want to rebase it and/or update the lsp tests?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/merlin/pull/919#issuecomment-466996584>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAB3gn90-5LA1xxOc-ElhSSH0W1KKDWkks5vQ9mngaJpZM4bDq_3>
> .
>
-- 
Andrey Popp / 8mayday@gmail.com

trefis commented:
I don't have what's needed to run the lsp tests, so I can't really help you on that part.
But since it seems likely that *I* actually broke them, what I'll do is extend the protocol a bit, and then write some tests in the usual mdx framework.

Hopefully once I debug issues there, the LSP ones should magically disappear.

andreypopp commented:
> I don't have what's needed to run the lsp tests, so I can't really help you on that part.
But since it seems likely that _I_ actually broke them, what I'll do is extend the protocol a bit, and then write some tests in the usual mdx framework.

No problem at all! I'll get to it soon — I'm interested in what broke this too.

Making this a part of original protocol is a very good idea as I can remove code from ocamlmerlin-lsp then and follow the theme of just converting LSP into original protocol there.

I can take a stub at it.

andreypopp commented:
@trefis pushed a fix — I think there was a type which restricted to `labels` namespace even if we didn't ask for a label. Do you agree?

Tests should pass now and it's ready to be merged.

trefis commented:
Yes that looks good. Thanks for cathing that!
